agent/exec/plugin/controller.go

```diff
@@ -0,0 +1,148 @@
+package plugin
+
+import (
+	engineapi "github.com/docker/docker/client"
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// controller implements agent.Controller against docker's API.
+//
+// Most operations against docker's API are done through the plugin name,
+// which is unique to the task.
+type controller struct {
+	task    *api.Task
+	adapter *pluginAdapter
+	closed  chan struct{}
+	err     error
+}
+
+var _ exec.Controller = &controller{}
+
+// NewController returns a docker exec controller for the provided task.
+func NewController(client engineapi.APIClient, task *api.Task) (exec.Controller, error) {
+	adapter, err := newPluginAdapter(client, task)
+	if err != nil {
+		return nil, err
+	}
+
+	return &controller{
+		task:    task,
+		adapter: adapter,
+		closed:  make(chan struct{}),
+	}, nil
+}
+
+func (r *controller) Task() (*api.Task, error) {
+	return r.task, nil
+}
+
+// Update tasks a recent task update and applies it to the plugin.
+func (r *controller) Update(ctx context.Context, t *api.Task) error {
+	log.G(ctx).Warnf("task updates not yet supported")
+	// TODO(stevvooe): While assignment of tasks is idempotent, we do allow
+	// updates of metadata, such as labelling, as well as any other properties
+	// that make sense.
+	return nil
+}
+
+// Prepare create a plugin and ensures the image is pulled.
+func (r *controller) Prepare(ctx context.Context) error {
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	return r.adapter.install(ctx)
+}
+
+// Enable the plugin.
+func (r *controller) Start(ctx context.Context) error {
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	if err := r.adapter.enable(ctx); err != nil {
+		return errors.Wrap(err, "enabling plugin failed")
+	}
+
+	return nil
+}
+
+// Wait forever.
+func (r *controller) Wait(pctx context.Context) error {
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	<-pctx.Done()
+	return pctx.Err()
+}
+
+// Shutdown the plugin cleanly.
+func (r *controller) Shutdown(ctx context.Context) error {
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	if err := r.adapter.disable(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Terminate the plugin.
+func (r *controller) Terminate(ctx context.Context) error {
```

Comment:
atm, disable does a SIGKILL. There's no graceful exit. I'll open a PR in docker/docker to address that. @vieux : Please add a TODO here so that it can be addressed accordingly, after the docker PR.